bindings/matrix-sdk-ffi/src/timeline.rs

```diff
@@ -0,0 +1,409 @@
+            VecDiff::Clear {} => TimelineChange::Clear,
+        }
+    }
+
+    pub fn replace(self: Arc<Self>) -> Option<Vec<Arc<TimelineItem>>> {
+        unwrap_or_clone_arc_into_variant!(self, .0, VecDiff::Replace { values } => values)
+    }
+
+    pub fn insert_at(self: Arc<Self>) -> Option<InsertAtData> {
+        unwrap_or_clone_arc_into_variant!(self, .0, VecDiff::InsertAt { index, value } => {
+            InsertAtData { index: index.try_into().unwrap(), item: value }
+        })
+    }
+
+    pub fn update_at(self: Arc<Self>) -> Option<UpdateAtData> {
+        unwrap_or_clone_arc_into_variant!(self, .0, VecDiff::UpdateAt { index, value } => {
+            UpdateAtData { index: index.try_into().unwrap(), item: value }
+        })
+    }
+
+    pub fn remove_at(&self) -> Option<u32> {
+        match &self.0 {
+            VecDiff::RemoveAt { index } => Some((*index).try_into().unwrap()),
+            _ => None,
+        }
+    }
+
+    pub fn r#move(&self) -> Option<MoveData> {
+        match &self.0 {
+            VecDiff::Move { old_index, new_index } => Some(MoveData {
+                old_index: (*old_index).try_into().unwrap(),
+                new_index: (*new_index).try_into().unwrap(),
+            }),
+            _ => None,
+        }
+    }
+
+    pub fn push(self: Arc<Self>) -> Option<Arc<TimelineItem>> {
+        unwrap_or_clone_arc_into_variant!(self, .0, VecDiff::Push { value } => value)
+    }
+}
+
+pub struct InsertAtData {
+    pub index: u32,
+    pub item: Arc<TimelineItem>,
+}
+
+pub struct UpdateAtData {
+    pub index: u32,
+    pub item: Arc<TimelineItem>,
+}
+
+pub struct MoveData {
+    pub old_index: u32,
+    pub new_index: u32,
+}
+
+#[derive(Clone, Copy)]
+pub enum TimelineChange {
+    Replace,
+    InsertAt,
+    UpdateAt,
+    RemoveAt,
+    Move,
+    Push,
+    Pop,
+    Clear,
+}
+
+#[repr(transparent)]
+#[derive(Clone)]
+pub struct TimelineItem(matrix_sdk::room::timeline::TimelineItem);
+
+impl TimelineItem {
+    fn from_arc(arc: Arc<matrix_sdk::room::timeline::TimelineItem>) -> Arc<Self> {
+        // SAFETY: This is valid because Self is a repr(transparent) wrapper
+        //         around the other Timeline type.
+        unsafe { Arc::from_raw(Arc::into_raw(arc) as _) }
+    }
+
+    pub fn as_event(self: Arc<Self>) -> Option<Arc<EventTimelineItem>> {
+        use matrix_sdk::room::timeline::TimelineItem as Item;
+        unwrap_or_clone_arc_into_variant!(self, .0, Item::Event(evt) => {
+            Arc::new(EventTimelineItem(evt))
+        })
+    }
+
+    pub fn as_virtual(self: Arc<Self>) -> Option<Arc<VirtualTimelineItem>> {
+        use matrix_sdk::room::timeline::TimelineItem as Item;
+        unwrap_or_clone_arc_into_variant!(self, .0, Item::Virtual(vt) => {
+            Arc::new(VirtualTimelineItem(vt))
+        })
+    }
+}
+
+#[derive(Clone)]
+pub struct EventTimelineItem(matrix_sdk::room::timeline::EventTimelineItem);
+
+impl EventTimelineItem {
```

Comment:
Last 2 thoughts for now - in EX, we're adding 2 more properties around these: `isOutgoing` (could maybe be named better but means was it sent by the current user) and `shouldShowSenderDetails` which is `false` if the previous item was sent by the same sender on the same day within a certain time threshold (again could probably be named better as its indicating a continual section of messages).

Not sure if these are too specific for the Rust side but mentioning as they would need to be re-implemented in EA too, so potentially worth considering?

Reply:
Those both sound worth adding 👍🏼

Reply:
Thinking about this, the `isOutgoing` property might also be a good place to include the transaction ID so its possible to identify when an event that is currently onscreen as a local echo gets replaced by the echo from the server.